stdlib/2/urllib2.pyi

```diff
@@ -45,14 +47,21 @@ class Request(object):
     def get_header(self, header_name: str, default: Optional[str] = ...) -> str: ...
     def header_items(self): ...
 
+class _Response(IO, metaclass=ABCMeta):
+    def geturl(self) -> str: ...
+    def info(self) -> Message: ...
+    def getcode(self) -> Optional[int]: ...
+
 class OpenerDirector(object):
+    addheaders = ...  # type: List[Tuple[str, str]]
+
     def add_handler(self, handler: BaseHandler) -> None: ...
-    def open(self, url: Union[Request, _string], data: Optional[_string] = ..., timeout: Optional[float] = ...): ...
+    def open(self, url: Union[Request, _string], data: Optional[_string] = ..., timeout: Optional[float] = ...) -> Optional[_Response]: ...
```

Comment:
The first argument is called `fullurl` (important when someone uses keyword argument syntax).

The `timeout` parameter is a bit problematic. It accepts float values and defaults to the singleton `socket._GLOBAL_DEFAULT_TIMEOUT`, which is just an `object` instance. But I don't think `None` is accepted, so it should not be `Optional`.

Reply:
I fixed `OpenerDirector.open(fullurl)` in 3039be30edf94b227d3c5512fc4ef771046a87f7. Thanks for your feedback.

About `timeout`: `urlib2` delegates to `urllib` delegates to `ftplib` and `httplib`, which both use `socket`. Quoting <https://docs.python.org/2/library/socket.html#socket.socket.settimeout>:
> ... The value argument can be a nonnegative float expressing seconds, or None. ... Setting a timeout of None disables timeouts on socket operations.
And yes, passing `None` to `urllib2.urlopen(timeout)` is passed all the way down to `socket.settimeout()` and disables the timeout. That may be a gory implementation detail, but is useful:
```python
import socket, urllib2
socket.setdefaulttimeout(0.01)
c = urllib2.urlopen('https://python.org/', timeout=None)
```
See commit message of 46d856847e5bcced97b974fa9060212e9cb71674 for more details.